cairo/ethereum/cancun/bloom.cairo

```diff
@@ -0,0 +1,167 @@
+from ethereum.utils.numeric import max, Uint_from_be_bytes, divmod, is_zero
+from ethereum.utils.bytes import Bytes20_to_Bytes, Bytes32_to_Bytes
+from src.utils.bytes import uint256_to_bytes32, felt_to_bytes16_little
+from src.utils.utils import Helpers
+from src.utils.dict import dict_squash
+
+from ethereum.crypto.hash import keccak256
+from ethereum.cancun.blocks import TupleLog
+from ethereum.cancun.fork_types import Bloom
+
+const BIT_MASK_11_BITS = 0x07FF;
+
+struct MutableBloomStruct {
+    dict_ptr_start: Bytes1DictAccess*,
+    dict_ptr: Bytes1DictAccess*,
+    len: felt,
+}
+
+struct MutableBloom {
+    value: MutableBloomStruct*,
+}
+
+func add_to_bloom{
+    range_check_ptr, bitwise_ptr: BitwiseBuiltin*, keccak_ptr: KeccakBuiltin*, bloom: MutableBloom
+}(bloom_entry: Bytes) {
+    alloc_locals;
+    let hash = keccak256(bloom_entry);
+    let (hash_low_bytes: felt*) = alloc();
+    felt_to_bytes16_little(hash_low_bytes, hash.value.low);
+    tempvar hash_bytes = Bytes(new BytesStruct(hash_low_bytes, 16));
+    _add_bloom_index(hash_bytes, 0);
+    _add_bloom_index(hash_bytes, 2);
+    _add_bloom_index(hash_bytes, 4);
+
+    return ();
+}
+
+// @dev does not expect to be called with indexes other than 0, 2, 4
+func _add_bloom_index{range_check_ptr, bitwise_ptr: BitwiseBuiltin*, bloom: MutableBloom}(
+    bloom_entry_hash: Bytes, index: felt
+) {
+    alloc_locals;
+    tempvar hash_subset = Bytes(new BytesStruct(bloom_entry_hash.value.data + index, 2));
+    let hash_subset_uint = Uint_from_be_bytes(hash_subset);
+
+    assert bitwise_ptr.x = hash_subset_uint.value;
+    assert bitwise_ptr.y = BIT_MASK_11_BITS;
+    tempvar bit_to_set = bitwise_ptr.x_and_y;
+    let bitwise_ptr = bitwise_ptr + BitwiseBuiltin.SIZE;
+
+    let bit_index = BIT_MASK_11_BITS - bit_to_set;
+    let (byte_index, bit_position) = divmod(bit_index, 8);
+
+    // Calculate 1 << (7 - bit_position) using multiplication
+    // This is equivalent to 2^bit_shift where bit_shift = 7 - bit_position
+    let (_, bit_shift_value) = divmod(7 - bit_position, 8);
+    let (bit_value) = Helpers.pow2(bit_shift_value);
+
+    // Read current byte value
+    let bloom_ptr = cast(bloom.value.dict_ptr, DictAccess*);
+    let (current_value) = dict_read{dict_ptr=bloom_ptr}(byte_index);
+
+    // Set the bit using OR operation
+    assert bitwise_ptr.x = current_value;
+    assert bitwise_ptr.y = bit_value;
+    let new_value = bitwise_ptr.x_or_y;
+    let bitwise_ptr = bitwise_ptr + BitwiseBuiltin.SIZE;
+
+    // Write back the updated byte
+    dict_write{dict_ptr=bloom_ptr}(byte_index, new_value);
+
+    tempvar bloom = MutableBloom(
+        new MutableBloomStruct(
+            dict_ptr_start=bloom.value.dict_ptr_start,
+            dict_ptr=cast(bloom_ptr, Bytes1DictAccess*),
+            len=bloom.value.len,
+        ),
+    );
+
+    return ();
+}
+
+func logs_bloom{range_check_ptr, bitwise_ptr: BitwiseBuiltin*, keccak_ptr: KeccakBuiltin*}(
```

Comment:
I wonder if logs bloom should be entirely done in a hint, and we simply assert that for each log in our list of logs, the log bloom has bit flipped to true for each log / topic?

not sure it changes a lot, except that we don't need a mutable bloom struct that accumulates the bloom

Reply:
we can start with this and check profiler
or ask @ClementWalter

Reply:
our codebase is not ready for this yet (as this would require significant work on the rust side). But we can keep it for later.

edit: Actually im thinking about how it would look like:

do the bit-flip operation in python, which would return a segment with the correct values. 
Then, iterate on this segment, verifying that: 
- for all topic/logs, the value at the given bit index is set
- for the rest of the values, the value is zero.

HOWEVER
in order to ensure all values are zero, we'd need to iterate another time on the segment, ensuring that un-touched bytes are zero, which would require tracking touched bytes, which would require a dict, so perhaps it's not that good in the end.

Reply:
since values are 1s or 0s, you could get away with:
- count the 1s, check that for all topic/logs, the value at the given bit index is set, check that the sum of all values in the segments are equal to the sum of the previously counted 1s?

Reply:
https://github.com/kkrt-labs/keth/issues/600